renpy.el

```diff
@@ -1529,6 +1529,109 @@ Uses `renpy-beginning-of-block', `renpy-end-of-block'."
   (renpy-end-of-block)
   (exchange-point-and-mark))
 
+;;;; Completion
+
+(defcustom renpy-setup-completion t
+  "Non-nil means Renpy mode sets up completion for the buffer."
+  :type 'boolean
+  :group 'renpy)
+
+(defun renpy--completion-context ()
+  "Return the completion category for point."
+  ;; TODO: Make sure the point is not within a comment or a string.
+  ;; TODO: More specific contexts: show text, show expression, ...
+  (save-excursion
+    (skip-syntax-backward "w_")
+    (let* ((_ (skip-syntax-backward " "))
+           (prev  (thing-at-point 'symbol)))
+      (cond
+       ;; call <point>
+       ((equal prev "call") :label)
+       ;; jump <point>
+       ((equal prev "jump") :label)
+       ;; show/scene/hide <point>
+       ((member prev '("show" "scene" "hide")) :image)
+       ;; at <point>
+       ((equal prev "at") :transform)
+       ;; default
+       (t :none)))))
+
+(defvar renpy--label-definition-re
+  (renpy-rx label-keyword (1+ space) (group label-name))
+  "Regexp for looking up label definitions.")
+
+(defun renpy--collect-labels ()
+  "Return all unique label names in the current buffer."
+  (let (labels)
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward renpy--label-definition-re nil t)
+        (push (match-string-no-properties 1) labels)))
+    labels))
+
+(defvar renpy--image-definition-re
+  (renpy-rx image-keyword (1+ space)
+	    (group image-name		; tag
+		   (0+ space name))	; attributes
+	    (0+ space) "=")
+  "Regexp for looking up image definitions.")
+
+(defun renpy--collect-images ()
+  "Return all unique image names in the current buffer."
+  (let (images)
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward renpy--image-definition-re nil t)
+        (push (match-string-no-properties 1) images)))
+    images))
+
+(defvar renpy--transform-definition-re
+  (renpy-rx transform-keyword (1+ space) (group name))
+  "Regexp for looking up transform definitions.")
+
+(defun renpy--collect-transforms ()
+  "Return all unique transform names in the current buffer."
+  (let (transforms)
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward renpy--transform-definition-re nil t)
+        (push (match-string-no-properties 1) transforms)))
+    transforms))
+
+;;;###autoload
+(defun renpy-completion-at-point ()
+  "Provide completion data for the symbol at point in Ren'Py buffers."
+  (when (derived-mode-p 'renpy-mode)
```

Comment:
I don't think an autoload is needed here because this isn't an entry point for using the mode. I don't think the major mode check is needed either, unless you had a use case where this is going to be called outside of renpy-mode?

Reply:
can't think of anything. I'll drop the check.